### PR TITLE
Extend Clojure BERT example

### PR DESCRIPTION
**Pull request by daveliepmann**
## Description ##
This PR extends the BERT sentence pair example in Clojure to include testing of the fine-tuned model on individual sentence pair samples. I discussed this last week with @gigasquid [on Slack](https://the-asf.slack.com/archives/CEE9X9WN7/p1558103162000500).

All the important changes are in the rich comment at the bottom of `bert_sentence_classification`, and are intended to be explored with a REPL. I plan to copy the REPL-driven example to the iPython notebook example after I'm sure my approach is correct but before merging.

I think my approach is correct but would like to double-check the following before merging:
 - [x] is there a way to pass the fine-tuned model directly to the `infer` API, rather than creating a factory over a saved checkpoint?
 - [x] is my interpretation of results correct? I included some individual samples that surprised me.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- ~~[ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)~~
- [x] Changes are complete (i.e. I finished coding on this PR)
~~- [ ] All changes have test coverage:~~
- [x] Code is well-documented: 
- For new examples,  ~~README.md is~~ *inline comments are* added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
 - [x] TODO copy the REPL-driven example to the iPython notebook

### Changes ###
- [x] evaluate Clojure BERT sentence pair example with individual sentence pair samples

## Comments ##

### CODE REVIEW

**Comment by karan6181** [comment]
@mxnet-label-bot add [Clojure, pr-work-in-progress]

**Comment by gigasquid** [comment]
Thanks @daveliepmann - Looks great so far!
Here are the answers to your questions:

_is there a way to pass the fine-tuned model directly to the infer API, rather than creating a factory over a saved checkpoint?_

No - there currently isn't a way to do that. It's a good idea to investigate :)

 _is my interpretation of results correct? I included some individual samples that surprised me._
A couple things to keep in mind:

1. The example is really on for demostration purposes. In the original Gluon NLP tutorial there is a Conclusion section https://gluon-nlp.mxnet.io/examples/sentence_embedding/bert.html

```For demonstration purpose, we skipped the warmup learning rate schedule and validation on dev dataset used in the original implementation. ```

We don't have a validation set only a training set that is going to affect the fine-tuning. We are also only running it for 3 epochs, we only have a training accuracy of about 0.70

2. Your results are going to be better the closer it is to your fine tuned data. Some of your made up sentences have words that might not be in the vocab as well. Any words that are not in there get assigned an unknown token.

In general, I think it's a great addition and I am happy to see it come along :)

File: contrib/clojure-package/examples/bert/fine-tune-bert.ipynb [hunk 1]
```diff
@@ -489,6 +492,122 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 22,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "[0.2633881 0.7366119]"
```
Comment (kedarbellare) [inline contrib/clojure-package/examples/bert/fine-tune-bert.ipynb]: what's this the output of?

File: contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj [hunk 2]
```diff
@@ -157,4 +183,42 @@
...
+                                                  (ndarray/array (slice-inputs-data processed-test-data 2) [1])])]
+      (ndarray/->vec (first prediction))))
+
+  ;; Modify an existing sentence pair to test:
+  ;; ["1"
+  ;;  "69773"
+  ;;  "69792"
+  ;;  "Cisco pared spending to compensate for sluggish sales ."
+  ;;  "In response to sluggish sales , Cisco pared spending ."]
+  (predict-equivalence fine-tuned-predictor
```
Comment (kedarbellare) [inline contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj]: did you want to add a test for this?

**Comment by kedarbellare** [comment]
@daveliepmann can you rebase your code? i think that should hopefully fix the ci failures. also i think there's a way to perform prediction without saving the model checkpoint but it's not well documented or straightforward AFAIK. it'll be good to have this (even if it's not typically used).

**Comment by Chouffe** [comment]
Will take a look this week @kedarbellare. Thanks for your contribution @daveliepmann :)
I am looking forward to reviewing it!

File: contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj [hunk 3]
```diff
@@ -138,14 +162,16 @@
...
-                                             :aux-params (m/aux-params bert-base)
-                                             :optimizer (optimizer/adam {:learning-rate 5e-6 :episilon 1e-9})
-                                             :batch-end-callback (callback/speedometer batch-size 1)})})))
+        fitted-model (m/fit (m/module model-sym {:contexts [dev]
+                                                 :data-names ["data0" "data1" "data2"]})
+                            {:train-data train-data  :num-epoch num-epoch
+                             :fit-params (m/fit-params {:allow-missing true
+                                                        :arg-params (m/arg-params bert-base)
+                                                        :aux-params (m/aux-params bert-base)
+                                                        :optimizer (optimizer/adam {:learning-rate 5e-6 :epsilon 1e-9})
```
Comment (Chouffe) [inline contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj]: Do we know why we pick this hyperparameters? Related to a paper?

File: contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj [hunk 4]
```diff
@@ -138,14 +162,16 @@
...
+                                                        :batch-end-callback (callback/speedometer batch-size 1)})})]
+    (m/save-checkpoint fitted-model {:prefix fine-tuned-prefix :epoch num-epoch})
```
Comment (Chouffe) [inline contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj]: Why do we save the model to disk now? Could we pass in a parameter to the function to to it? This function seems to do too many things?

File: contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj [hunk 5]
```diff
@@ -157,4 +183,42 @@
...
-  (m/save-checkpoint model {:prefix "fine-tune-sentence-bert" :epoch 3}))
+
+  (m/save-checkpoint model {:prefix fine-tuned-prefix :epoch 3})
+
+  
+  ;;;; Explore results from the fine-tuned model
+
+  ;; We need a predictor with a batch size of 1, so we can feed the
+  ;; model a single sentence pair.
+  (def fine-tuned-predictor
```
Comment (Chouffe) [inline contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj]: Nice!!

File: contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj [hunk 6]
```diff
@@ -157,4 +183,42 @@
...
+    (infer/create-predictor (infer/model-factory fine-tuned-prefix
+                                                 [{:name "data0" :shape [1 seq-length] :dtype dtype/FLOAT32 :layout layout/NT}
+                                                  {:name "data1" :shape [1 seq-length] :dtype dtype/FLOAT32 :layout layout/NT}
+                                                  {:name "data2" :shape [1]            :dtype dtype/FLOAT32 :layout layout/N}])
+                            {:epoch 3}))
```
Comment (Chouffe) [inline contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj]: Why do we need this hardcoded epoch number here?

File: contrib/clojure-package/examples/infer/predictor/src/infer/predictor_example.clj [hunk 7]
```diff
@@ -99,3 +99,9 @@
       (:help options) (println summary)
       (some? errors) (println (join "\n" errors))
       :else (run-predictor options))))
+
+(comment
+  (run-predictor {:model-path-prefix "models/resnet-18/resnet-18"
```
Comment (Chouffe) [inline contrib/clojure-package/examples/infer/predictor/src/infer/predictor_example.clj]: This seems to be unrelated to the BERT work. Should it be done in another PR?

File: contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj [hunk 8]
```diff
@@ -16,42 +16,67 @@
 ;;
 
 (ns bert.bert-sentence-classification
+  "Fine-tuning Sentence Pair Classification with BERT
```
Comment (Chouffe) [inline contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj]: Awesome docstring! Thanks a lot!

File: contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj (hunk 3 above)
Comment (daveliepmann) [reply contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj]: These were all pre-existing in the Clojure example, which AFAIK were in the non-Clojure example this was based on, which I assume were from the BERT paper but have no idea. :)

File: contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj (hunk 4 above)
Comment (daveliepmann) [reply contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj]: We have to save the model to disk because there's no other way but saving to disk and loading it back in to get a prediction out of it.

File: contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj (hunk 6 above)
Comment (daveliepmann) [reply contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj]: As I recall, we have to hard-code the epoch because otherwise we don't know which saved model to load from disk.

File: contrib/clojure-package/examples/infer/predictor/src/infer/predictor_example.clj (hunk 7 above)
Comment (daveliepmann) [reply contrib/clojure-package/examples/infer/predictor/src/infer/predictor_example.clj]: It is related only because it came up while exploring ways to predict with models that aren't saved from disk. I left it in because otherwise one doesn't have a way to run this code from a REPL. I can remove it from this PR but I don't expect I have time right now to shepherd another PR. Would you like me to do that?

File: contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj (hunk 6 above)
Comment (Chouffe) [reply contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj]: With the `infer` API I suppose? Maybe we could change this at some point @kedarbellare?

File: contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj (hunk 4 above)
Comment (Chouffe) [reply contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj]: With the `infer` API I suppose? Maybe we could change this at some point @kedarbellare?

File: contrib/clojure-package/examples/infer/predictor/src/infer/predictor_example.clj (hunk 7 above)
Comment (Chouffe) [reply contrib/clojure-package/examples/infer/predictor/src/infer/predictor_example.clj]: It should be fine I think

File: contrib/clojure-package/examples/bert/fine-tune-bert.ipynb (hunk 1 above)
Comment (daveliepmann) [reply contrib/clojure-package/examples/bert/fine-tune-bert.ipynb]: `[0.2633881 0.7366119]` is the output of:

```clojure
(predict-equivalence fine-tuned-predictor
                     "The company cut spending to compensate for weak sales ."
                     "In response to poor sales results, the company cut spending .")
```
I'm not sure why the result appears before its expression in the *.ipynb* file, but on my machine it displays this pair correctly as "In [22]" followed by "Out [22]".

File: contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj (hunk 2 above)
Comment (daveliepmann) [reply contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj]: Can do

Comment (daveliepmann) [reply contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj]: Just pushed one. Thanks for the idea—my PR broke the existing test, but I guess tests for examples aren't part of the CI checks.

File: contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj (hunk 6 above)
Comment (daveliepmann) [reply contrib/clojure-package/examples/bert/src/bert/bert_sentence_classification.clj]: Responding to edit: `num-epoch` isn't in scope in the `comment`. I decided against defining it globally in order to parameterize a short REPL exploration. 

Another reason not to `def` the `3` here is that `num-epoch` is a value meant to be passed in from the command line, and the rich comment code is a parallel of invoking from the command line with that argument. So at a minimum we would need a new name.

**Comment by kedarbellare** [comment]
thanks @daveliepmann!! :100:
